scanpy/preprocessing/_simple.py

```diff
@@ -743,59 +743,127 @@ def scale(
+    axis
+        Sample observations (axis=0) or variables (axis=1). Default is 0.
 
     Returns
     -------
-    Returns `X[obs_indices], obs_indices` if data is array-like, otherwise
-    subsamples the passed :class:`~anndata.AnnData` (`copy == False`) or
-    returns a subsampled copy of it (`copy == True`).
+    Returns `X[indices] or X[:, indices], indices` depending on the axis
+    argument if data is array-like, otherwise samples the passed
+    :class:`~anndata.AnnData` (`copy == False`) or returns a sampled
+    copy of it (`copy == True`).
     """
     np.random.seed(random_state)
-    old_n_obs = data.n_obs if isinstance(data, AnnData) else data.shape[0]
-    if n_obs is not None:
-        new_n_obs = n_obs
+    old_n = data.shape[axis]
+    if axis not in (0, 1):
+        raise ValueError(f'`axis` must be either 0 or 1.')
+    if fraction is None and n is None:
+        raise ValueError(f'Either `fraction` or `n` must be set.')
+    if fraction is not None and n is not None:
+        raise ValueError(f'Providing both `fraction` and `n` is not allowed.')
+    if n is not None:
+        new_n = n
     elif fraction is not None:
-        if fraction > 1 or fraction < 0:
+        if fraction < 0:
+            raise ValueError(f'`fraction needs to be nonnegative`, not {fraction}')
+        if not replace and fraction > 1:
             raise ValueError(
-                f'`fraction` needs to be within [0, 1], not {fraction}'
+                f'If replace=False, `fraction` needs to be within [0, 1], not {fraction}'
             )
-        new_n_obs = int(fraction * old_n_obs)
-        logg.debug(f'... subsampled to {new_n_obs} data points')
+        new_n = int(fraction * old_n)
+        obs_or_var_str = 'observations' if axis == 0 else 'variables'
+        logg.debug(f'... sampled to {new_n} {obs_or_var_str}')
     else:
         raise ValueError('Either pass `n_obs` or `fraction`.')
-    obs_indices = np.random.choice(old_n_obs, size=new_n_obs, replace=False)
+    indices = np.random.choice(old_n, size=new_n, replace=replace)
     if isinstance(data, AnnData):
         if copy:
-            return data[obs_indices].copy()
+            view = data[indices] if axis == 0 else data[:, indices]
+            return view.copy()
         else:
-            data._inplace_subset_obs(obs_indices)
+            if axis == 0:
+                data._inplace_subset_obs(indices)
+            else:
+                data._inplace_subset_var(indices)
     else:
         X = data
-        return X[obs_indices], obs_indices
+        return X[indices] if axis == 0 else X[:, indices], indices
+
+
+def subsample(
+    data: Union[AnnData, np.ndarray, spmatrix],
+    fraction: Optional[float] = None,
+    n_obs: Optional[int] = None,
+    random_state: AnyRandom = 0,
+    copy: bool = False,
+) -> Optional[AnnData]:
+    """\
+    Subsample to a fraction of the number of observations.
+
+    .. warning::
+        .. deprecated:: 1.4.7
```

Comment:
Should this function also throw a deprecation function on use?

Reply:
There are a few if you do a full code search for `depreca`.

One example would be at the top of `calculate_qc_metrics`. In anndata we also have a whole file for testing that deprecated functions work and throw warnings, which you could look at for examples of tests.

Reply:
handled in #3380